contracts/upgradeable_contracts/U_ForeignBridge.sol

```diff
@@ -110,30 +108,32 @@ contract ForeignBridge is ERC677Receiver, Validatable {
         return IBurnableMintableERC677Token(addressStorage[keccak256("erc677token")]);
     }
 
-    function setGasLimits(uint256 _gasLimitDepositRelay, uint256 _gasLimitWithdrawConfirm) public onlyOwner {
+    function setGasLimits(uint256 _gasLimitDepositRelay, uint256 _gasLimitWithdrawConfirm) external onlyOwner {
         uintStorage[keccak256("gasLimitDepositRelay")] = _gasLimitDepositRelay;
         uintStorage[keccak256("gasLimitWithdrawConfirm")] = _gasLimitWithdrawConfirm;
         emit GasConsumptionLimitsUpdated(gasLimitDepositRelay(), gasLimitWithdrawConfirm());
     }
 
-    function deposit(address recipient, uint256 value, bytes32 transactionHash) public onlyValidator {
-        require(address(erc677token()) != address(0x0));
-
-        // Protection from misbehaing authority
+    function deposit(address recipient, uint256 value, bytes32 transactionHash) external onlyValidator {
         bytes32 hashMsg = keccak256(recipient, value, transactionHash);
         bytes32 hashSender = keccak256(msg.sender, hashMsg);
         // Duplicated deposits
         require(!depositsSigned(hashSender));
         setDepositsSigned(hashSender, true);
 
-        uint256 signed = numDepositsSigned(hashMsg).add(1);
+        uint256 signed = numDepositsSigned(hashMsg);
+        require(!isAlreadyProcessed(signed));
+        // the check above assumes that the case when the value could be overflew will not happen in the addition operation below
+        signed = signed + 1;
+
         setNumDepositsSigned(hashMsg, signed);
 
         emit SignedForDeposit(msg.sender, transactionHash);
 
         if (signed == validatorContract().requiredSignatures()) {
```

Comment:
There is another issue that concerns me. What if `requiredSignatures` is decreased below the current `signed` value for some reason? Current deposit would never be processed.
Looks like `>=` should be used instead of `==`.